applications/RomApplication/custom_strategies/rom_builder_and_solver.h

```diff
@@ -158,37 +158,55 @@ class ROMBuilderAndSolver : public BuilderAndSolver<TSparseSpace, TDenseSpace, T
         set_type dof_global_set;
         dof_global_set.reserve(number_of_elements * 20);
 
-        #pragma omp parallel firstprivate(dof_list, second_dof_list)
+        double NumberOfHRomElements=0.0;
+        #pragma omp parallel firstprivate(dof_list, second_dof_list) reduction(+:NumberOfHRomElements)
         {
-            auto &r_current_process_info = rModelPart.GetProcessInfo();
+            const ProcessInfo& r_current_process_info = rModelPart.GetProcessInfo();
 
             // We cleate the temporal set and we reserve some space on them
             set_type dofs_tmp_set;
             dofs_tmp_set.reserve(20000);
-
             // Gets the array of elements from the modeler
             #pragma omp for schedule(guided, 512) nowait
             for (int i = 0; i < number_of_elements; ++i)
             {
                 auto it_elem = r_elements_array.begin() + i;
-
+                //detect whether the element has a Hyperreduced Weight (H-ROM simulation) or not (ROM simulation)
+                if ((it_elem)->Has(HROM_WEIGHT))
+                    NumberOfHRomElements++;
+                else
+                    it_elem->SetValue(HROM_WEIGHT, 1.0);
                 // Gets list of Dof involved on every element
-                pScheme->GetElementalDofList(*(it_elem.base()), dof_list, r_current_process_info);
+                pScheme->GetDofList(*it_elem, dof_list, r_current_process_info);
                 dofs_tmp_set.insert(dof_list.begin(), dof_list.end());
             }
 
             // Gets the array of conditions from the modeler
             ConditionsArrayType &r_conditions_array = rModelPart.Conditions();
             const int number_of_conditions = static_cast<int>(r_conditions_array.size());
+
+            ModelPart::ConditionsContainerType mSelectedConditions_private;
             #pragma omp for schedule(guided, 512) nowait
             for (int i = 0; i < number_of_conditions; ++i)
             {
-                auto it_cond = r_conditions_array.begin() + i;
-
+                auto it_cond = r_conditions_array.ptr_begin() + i;

```

Comment:
do begin() instead of ptr_begin() and you will be able to remove a lot of asterisks in the following lines (it will also be faster, since you will not create a lot of smart pointers)